python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -0,0 +1,861 @@
+        # to datetime.
+        # assert all(
+        #     c.obj.type() == dtype
+        #     for c, dtype in zip(df.columns, self.schema.values())
+        # )
+        if self.predicate is None:
+            return df
+        else:
+            mask = self.predicate.evaluate(df)
+            return df.filter(mask)
+
+
+@dataclass(slots=True)
+class Cache(IR):
+    """
+    Return a cached plan node.
+
+    Used for CSE at the plan level.
+    """
+
+    key: int
+    """The cache key."""
+    value: IR
+    """The unevaluated node to cache."""
+
+    def evaluate(self, *, cache: dict[int, DataFrame]) -> DataFrame:
+        """Evaluate and return a dataframe."""
+        try:
+            return cache[self.key]
+        except KeyError:
+            return cache.setdefault(self.key, self.value.evaluate(cache=cache))
+
+
+@dataclass(slots=True)
+class DataFrameScan(IR):
+    """
+    Input from an existing polars DataFrame.
+
+    This typically arises from ``q.collect().lazy()``
+    """
+
+    df: Any
+    """Polars LazyFrame object."""
+    projection: list[str]
+    """List of columns to project out."""
+    predicate: Expr | None
+    """Mask to apply."""
+
+    def evaluate(self, *, cache: dict[int, DataFrame]) -> DataFrame:
+        """Evaluate and return a dataframe."""
+        pdf = pl.DataFrame._from_pydf(self.df)
+        if self.projection is not None:
+            pdf = pdf.select(self.projection)
+        # TODO: goes away when libcudf supports large strings
+        table = pdf.to_arrow()
+        schema = table.schema
+        for i, field in enumerate(schema):
+            if field.type == pa.large_string():
+                # TODO: Nested types
+                schema = schema.set(i, pa.field(field.name, pa.string()))
+        table = table.cast(schema)
+        df = DataFrame.from_table(
+            plc.interop.from_arrow(table), list(self.schema.keys())
+        )
+        assert all(
+            c.obj.type() == dtype for c, dtype in zip(df.columns, self.schema.values())
+        )
+        if self.predicate is not None:
+            mask = self.predicate.evaluate(df)
+            return df.filter(mask)
+        else:
+            return df
+
+
+@dataclass(slots=True)
+class Select(IR):
+    """Produce a new dataframe selecting given expressions from an input."""
+
+    df: IR
+    """Input dataframe."""
+    cse: list[Expr]
+    """
+    List of common subexpressions that will appear in the selected expressions.
+
+    These must be evaluated before the returned expressions.
+    """
+    expr: list[Expr]
+    """List of expressions to evaluate to form the new dataframe."""
+
+    def evaluate(self, *, cache: dict[int, DataFrame]):
+        """Evaluate and return a dataframe."""
+        df = self.df.evaluate(cache=cache)
+        for e in self.cse:
+            df = df.with_columns([e.evaluate(df)])
+        return DataFrame([e.evaluate(df) for e in self.expr], [])
+
+
+def placeholder_column(n: int):
```

Comment:
This is a libcudf thing. I want to make an `EMPTY` column of a given size (so it has no backing data), because the `count` aggregation (with `null_policy = INCLUDE_NULLS`) doesn't look at the data. But nothing in libcudf works with columns with dtype `EMPTY`.

Reply:
Maybe add a comment to #12477. At some point we'll have to make a call on that issue, and it would be good to know if empty is actually important. pylibcudf certainly exposes some cases where we end up out of step with pyarrow by not having a null type. I'm not sure if we care enough about them though.